### ISSUE TITLE
OOM （RAM not gpu） using DDP trainning

### ISSUE THREAD

**Issue by wcyjerry** [ultralytics/ultralytics#19671]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

Hi, 
    sorry to bother again, But I'm facing a OOM problem in multi-gpu training. See this:

![Image](https://github.com/user-attachments/assets/3048ac5c-2c4e-453f-9398-323281e58e34)

This happens less when I using less num_workers, while still may happen. And the main reason I have researched is that during the training process,  cache buffer is continuous increasing, until it consuming all ram, and the process get killed. see this:

![Image](https://github.com/user-attachments/assets/435c985e-c9fd-449b-9530-74281a46c16c)
I'm not sure why this happen, and whether it is a bug here. 
The version I'm using is 8.2.70, is there any updates to fix this in later version? 
I have modified the code so it is a huge work for me to always using the latest version, if it is not necessary. 
PLZ check this out for me, thanks man!

### Environment

Ultralytics YOLOv8.2.70 🚀 Python-3.8.10 torch-2.3.1+cu118 CUDA:0 (NVIDIA GeForce RTX 4090, 24217MiB)
Setup complete ✅ (112 CPUs, 251.5 GB RAM, 813.4/898.9 GB disk)

OS                  Linux-5.15.0-107-generic-x86_64-with-glibc2.17
Environment         Linux
Python              3.8.10
Install             git
RAM                 251.52 GB
CPU                 Intel Xeon Gold 6348 2.60GHz
CUDA                11.8

numpy               ✅ 1.24.4<2.0.0,>=1.23.0
matplotlib          ✅ 3.4.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 9.0.0>=7.1.2
pyyaml              ✅ 5.4.1>=5.3.1
requests            ✅ 2.25.1>=2.23.0
scipy               ✅ 1.10.1>=1.4.1
torch               ✅ 2.3.1+cu118>=1.8.0
torchvision         ✅ 0.18.1+cu118>=0.9.0
tqdm                ✅ 4.66.4>=4.64.0
psutil              ✅ 6.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 1.3.1>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.0>=2.0.0

### Minimal Reproducible Example

No

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @wcyjerry, thank you for reporting this issue and for your detailed description and environment details! 🚀 We appreciate your effort in bringing this to our attention. An Ultralytics engineer will review your report and assist you soon.

In the meantime, here are a few suggestions to help us investigate further:

- If this is a 🐛 Bug Report, please provide a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) that demonstrates the issue. This is crucial for us to debug the problem effectively.
- Ensure you are using the latest `ultralytics` package, as updates may have already resolved this issue. You can upgrade with the following command:

```bash
pip install -U ultralytics
```

- If you’ve modified the code, please test whether the issue persists with the unmodified version to help isolate the cause.

For additional resources:
- Visit the [Docs](https://docs.ultralytics.com/) for detailed guidance.
- Join our vibrant community! For real-time support, check out [Discord](https://discord.com/invite/ultralytics) 🎧, engage in discussions on [Discourse](https://community.ultralytics.com/), or visit our [Subreddit](https://reddit.com/r/Ultralytics).

## Verified Environments

YOLO can be run in any of these up-to-date environments (with dependencies preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

This badge indicates the status of our [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests that run daily and on every commit to verify the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across macOS, Windows, and Ubuntu.

Feel free to follow up with additional details or questions. We're here to help! 😊

**Comment by Y-T-G** [comment]
What's the training command used?

**Comment by wcyjerry** [comment]
> What's the training command used?
just normal:
![Image](https://github.com/user-attachments/assets/f1d26db0-94fd-4cb9-860a-38382ce7b3a1)

But I think I found the reason, seems like caused by pin_memory=True, when memory buffer is fulled, this pin_memory can't be free, cause the system kill the training process. But, the huge mount of memory usage is still not reasonable.
I tried set pin_memory=False in 

![Image](https://github.com/user-attachments/assets/6438430c-193c-41f8-9b44-0ccc95260c32)
and not get killed.

**Comment by glenn-jocher** [comment]
@wcyjerry thank you for sharing these details. Based on your findings, this appears to be a memory management issue in DDP training where pinned memory isn't being released properly. While your workaround to set `pin_memory=False` is valid, we recommend:

1. **Update to the latest YOLO version** (current is 8.2.70+) where we've made memory optimizations
2. Try reducing `workers` further (start with `workers=4`) 
3. Add `self._clear_memory()` calls in the training loop if using custom code

The team is actively working on DDP memory optimizations. If the issue persists after updating, please share a minimal reproducible example using our latest codebase.

**Comment by Minxiangliu** [comment]
Hi @glenn-jocher ,
Hello, may I ask how I can add the `self._clear_memory()` method if I use custom training? If possible, I would appreciate some suggestions. Thanks in advance!
```py
model = YOLO(model='pre-trained-models/yolo11m.pt', task='detect')
model.train(
    data='data.yaml', 
    epochs=500, 
    imgsz=640, 
    workers=32, 
    device=[0,1], 
    cache='disk', 
    batch=128, 
    patience=150)
```

**Comment by wcyjerry** [comment]
@glenn-jocher Hi, thanks for reply.
I have do some comparisons between latest code and mine. 

![Image](https://github.com/user-attachments/assets/2e8cf7f7-411a-438f-a559-20e2aa7a8388)

![Image](https://github.com/user-attachments/assets/cdb34bb9-5044-482b-994f-e5f4c4af1c95)
seems like in v8.27.0(my version) doesn't include clear_memory(). I will add this and try again.
BTW this is DDP's problem not ultralytics framework's? So any framework using ddp may meet this problem?

**Comment by wcyjerry** [comment]
@glenn-jocher Thanks for your advice.
1. I will update my version to latest one, but maybe needs some work, as I have developed a lot.
2. Yes, smaller num_workers wil help, but I'm training small model like v5n  v5s, the main bottleneck is io. So it's better to find some way to using larger nw.
3. Using official code.
Maybe there is something more beyond what I've observed, But my current solution is close pin_memory equals True. and GPT says it's reasonable:

![Image](https://github.com/user-attachments/assets/21e3fc3e-d4e7-46a0-a654-b4dfdd697b78)

**Comment by glenn-jocher** [comment]
To add memory clearing in custom training, you can create a custom callback that triggers `self._clear_memory()` after each epoch. Here's a minimal example using your existing code:

```python
from ultralytics.engine.trainer import BaseTrainer

class CustomTrainer(BaseTrainer):
    def on_fit_epoch_end(self):
        super().on_fit_epoch_end()
        if self._get_memory(fraction=True) > 0.8:  # Clear at 80% memory usage
            self._clear_memory()

model = YOLO('pre-trained-models/yolo11m.pt', task='detect')
model.train(trainer=CustomTrainer, ...)  # Keep your existing args
```

Your current solution of `pin_memory=False` is valid for reducing RAM pressure. The latest versions have additional memory optimizations that might help when you're able to upgrade.

**Comment by TNPIsHere** [comment]
> To add memory clearing in custom training, you can create a custom callback that triggers `self._clear_memory()` after each epoch. Here's a minimal example using your existing code:
> 
> from ultralytics.engine.trainer import BaseTrainer
> 
> class CustomTrainer(BaseTrainer):
>     def on_fit_epoch_end(self):
>         super().on_fit_epoch_end()
>         if self._get_memory(fraction=True) > 0.8:  # Clear at 80% memory usage
>             self._clear_memory()
> 
> model = YOLO('pre-trained-models/yolo11m.pt', task='detect')
> model.train(trainer=CustomTrainer, ...)  # Keep your existing args
> Your current solution of `pin_memory=False` is valid for reducing RAM pressure. The latest versions have additional memory optimizations that might help when you're able to upgrade.

Thank you! Is there any good reason not to clear it on every epoch?

**Comment by glenn-jocher** [comment]
Yes—clearing every epoch forces CUDA synchronizations and frequent allocator re-allocations, which usually slows DDP throughput; better to clear only when usage crosses a threshold or right before validation; for setup tips see the Multi‑GPU training guide in the docs. [Multi‑GPU training](https://docs.ultralytics.com/modes/train/#multi-gpu-training)

**Comment by TNPIsHere** [comment]
@glenn-jocher Thank you for responding. My issue is very similar to the topicstarter's.
I train lightweight yolov11s model for a single class detection hometask on a laptop (MacBook M4 Pro, 24 GB RAM), and seeing the same - on every epoch additional 300-500 MB of RAM is occupied by something.

I looked at the code and (of course) pin_memory is False by default on Mac OS. 
I also used your function above to clean memory (and as I see from the training script it works with Apple's MPS just fine). 
What I also tried is few versions of ultralytics between 8.3.0 to 8.3.180 because in another thread it helped someone to get rid of VRAM leakage (on NVIDIA's GPU though). 
I do not use image caching, I tried to lower the nominal batch size to minimize any excessive RAM usage - no success so far.

Probably, need to profile the training procedure on RAM usage carefully and inspect what's happening. 
BTW, I know it is not a good practice in production but I do the training in anacondas jupyter notebook - could this be somehow an issue?
This issue is really annoying as it does not allow to beat the threshold on mAP50-95 as it naturally limits the batchsize/epochs/resolution :^).

Would be happy to hear any comments/suggestions/recommendations from you.

**Comment by glenn-jocher** [comment]
Thanks for the detailed context—yes, Jupyter can retain references across cells (especially after upgrades), so please restart the kernel and also try an isolated CLI run to rule the notebook out: 

```bash
yolo detect train model=yolo11s.pt data=path/to/data.yaml device=mps workers=0 batch=16 epochs=2 save=False
```

If RAM still grows per epoch, please open a new issue and include the `yolo checks` output (or in‑notebook `!yolo checks`), the exact command you ran, text logs instead of screenshots, and the result of:

```python
from ultralytics.utils.checks import collect_system_info
print(collect_system_info())
```

For platform notes and known tips on Apple hardware, see the Apple Silicon (MPS) section of the training docs: [Apple Silicon (MPS) training guide](https://docs.ultralytics.com/modes/train/#apple-silicon-mps-training).
